Nlx unique chan id fix

sin-mike opened the pull request:
@samuelgarcia 

This PR includes a fix for non-unique chan_id when there are several files for the same AD channel

see  #590,  #589

Now I can read mixed records without splitting folders, like this
![image](https://user-images.githubusercontent.com/4338049/47603586-b3693980-d9f6-11e8-9380-f7a2df0bcb0e.png)

These two channels are reading from one AD channel, but with different preprocessings:
![image](https://user-images.githubusercontent.com/4338049/47603594-dbf13380-d9f6-11e8-8de2-ec6b4033c6c2.png)

pep8speaks commented:
Hello @sin-mike! Thanks for submitting the PR.

 - In the file [`neo/rawio/neuralynxrawio.py`](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py), following                are the PEP8 issues :
> [Line 71:38](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L71): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 77:29](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L77): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 78:33](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L78): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 78:61](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L78): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 88:88](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L88): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 95:9](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L95): [E731](https://duckduckgo.com/?q=pep8%20E731) do not assign a lambda expression, use a def
> [Line 95:27](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L95): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 95:29](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L95): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 102:13](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L102): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 102:42](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L102): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 113:1](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L113): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 126:21](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L126): [E266](https://duckduckgo.com/?q=pep8%20E266) too many leading '#' for block comment
> [Line 129:21](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L129): [E266](https://duckduckgo.com/?q=pep8%20E266) too many leading '#' for block comment
> [Line 170:50](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L170): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 511:27](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L511): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 521:1](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L521): [E266](https://duckduckgo.com/?q=pep8%20E266) too many leading '#' for block comment
> [Line 523:11](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L523): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 523:20](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L523): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 525:13](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L525): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 525:23](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L525): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 527:10](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L527): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 528:55](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L528): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 531:1](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L531): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 531:10](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L531): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\w'
> [Line 531:14](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L531): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 590:23](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L590): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\S'
> [Line 607:10](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L607): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\w'
> [Line 614:10](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L614): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\w'
> [Line 627:10](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L627): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\S'
> [Line 635:10](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L635): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\w'
> [Line 645:25](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L645): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\('
> [Line 645:32](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L645): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\)'
> [Line 645:45](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L645): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\S'
> [Line 645:51](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L645): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\('
> [Line 645:58](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L645): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\.'
> [Line 645:62](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L645): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\)'
> [Line 645:74](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L645): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\S'
> [Line 646:25](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L646): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\('
> [Line 646:32](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L646): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\)'
> [Line 646:45](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L646): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\S'
> [Line 646:51](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L646): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\('
> [Line 646:58](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L646): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\.'
> [Line 646:62](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L646): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\)'
> [Line 646:74](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L646): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\S'
> [Line 647:36](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L647): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\S'
> [Line 650:32](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L650): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\S'
> [Line 650:46](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L650): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\S'
> [Line 651:31](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L651): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\S'
> [Line 651:45](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L651): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\S'
> [Line 652:43](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L652): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\S'
> [Line 710:1](https://github.com/sin-mike/python-neo/blob/dc56f849b15d3a18c815b15d5c52372d612d8b14/neo/rawio/neuralynxrawio.py#L710): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

sin-mike commented:
Hi, @samuelgarcia, any updates on this PR?

sin-mike commented:
> 
> 
> @sin-mike
> Thank you very much for this. This is cool to fix.
> I was am a bit busy and still be until wensday.
> I will make comment soon.
> 
> For short, I am a bit discomfort with this uuid, this make the logic harder to understand.
> In my mind the double key (name, chan_id) was already unique for the dict.

@samuelgarcia Indeed, but here I was worried about the consistency, as, in the header, you're using just integer ids. And that using tuple ids could easily break something further. As I discovered, the header is a key structure for the data IO operations, e.g. to restrict IOs to certain channels, I just modified the header and removed rows partially. Therefore I used just integer IDs.

samuelgarcia commented:
@sin-mike I am back.
thank you for this new comit but I had in mind the posite solution : remove this generated uid.
I think we can go wihtout it. 

In short in the header['signal_channels'] the chan_id must totally consistent with what the user known about the file. So we must put the real chan_id even it is not unique.
Some IOs do not have unique id or unique names for sig channels. it is not problem because we can always  ```acces then read_signal_chunk(channel_indexes=[...])```. In your special case ```read_signal_chunk(channel_names=[...])``` and ```read_signal_chunk(channel_ids=[...])``` will raise an error because it is tested at the BaseRawIO class level.

Other remark: why did you change the id for spike channel ?

I will make comment in the code.

If we use a generated id it should be internal but in our case I think it is useless.

I feel really sorry to add more work.

samuelgarcia commented on neo/rawio/neuralynxrawio.py:
```diff
@@ -65,31 +65,41 @@ def _parse_header(self):
 
-        self.ncs_filenames = OrderedDict()  # chan_id: filename
-        self.nse_ntt_filenames = OrderedDict()  # chan_id: filename
+        self.ncs_filenames = OrderedDict()  # uid: filename
```
We can do this with a tupple for the key:
self.ncs_filenames = OrderedDict()  # (chan_name, chan_id): filename

samuelgarcia commented on neo/rawio/neuralynxrawio.py:
```diff
@@ -65,31 +65,41 @@ def _parse_header(self):
-        self.ncs_filenames = OrderedDict()  # chan_id: filename
-        self.nse_ntt_filenames = OrderedDict()  # chan_id: filename
+        self.ncs_filenames = OrderedDict()  # uid: filename
+        self.nse_ntt_filenames = OrderedDict()  # uid: filename
```
I prefer also :
 # (chan_name, chan_id): filename

samuelgarcia commented on neo/rawio/neuralynxrawio.py:
```diff
@@ -65,31 +65,41 @@ def _parse_header(self):
+        self.ncs_filenames = OrderedDict()  # uid: filename
+        self.nse_ntt_filenames = OrderedDict()  # uid: filename
         self.nev_filenames = OrderedDict()  # chan_id: filename
+        self.uid2chid = OrderedDict() # uid : (file_enum, chan_enum, chan_id)
```
I prefer also :
 (chan_name, chan_id): filename

samuelgarcia commented on neo/rawio/neuralynxrawio.py:
```diff
@@ -65,31 +65,41 @@ def _parse_header(self):
+        # filter all filenames to match IO extensions,
+        # so that fid indicates only the necessary files
+        all_filenames = filter(lambda f: os.path.splitext(f)[1][1:] in self.extensions, 
+                               sorted(os.listdir(self.dirname)))
```
Super good.

samuelgarcia commented on neo/rawio/neuralynxrawio.py:
```diff
@@ -65,31 +65,41 @@ def _parse_header(self):
+        # // Mike
+        gen_uid = lambda x,y,z: x + 1000*y
+
+        for fid, filename in enumerate(all_filenames):
```
I am quite unhappy with this id.

samuelgarcia commented on neo/rawio/neuralynxrawio.py:
```diff
@@ -98,19 +108,27 @@ def _parse_header(self):
+                
+                uid = gen_uid(fid, idx, chan_id)
+                self.uid2chid[uid] = (fid, idx, chan_id)
+
```
This could be remove so.

samuelgarcia commented on neo/rawio/neuralynxrawio.py:
```diff
@@ -98,19 +108,27 @@ def _parse_header(self):
-                    sig_channels.append((chan_name, chan_id, info['sampling_rate'],
+
+                    ### Note: vvvv: this structure could contain non-unique chan_name, chan_id
+                    sig_channels.append((chan_name, uid, info['sampling_rate'],
```
Let s go back to
sig_channels.append((chan_name, chan_id, info['sampling_rate'],
so the use can have something that match the real id in the file and the software.

samuelgarcia commented on neo/rawio/neuralynxrawio.py:
```diff
@@ -98,19 +108,27 @@ def _parse_header(self):
-                    self.ncs_filenames[chan_id] = filename
+                    ### // Mike
+
+                    self.ncs_filenames[uid] = filename
```
self.ncs_filenames[( chan_name, chan_id)] = filename

samuelgarcia commented on neo/rawio/neuralynxrawio.py:
```diff
@@ -133,26 +151,36 @@ def _parse_header(self):
+                    # assert chan_id not in self.nse_ntt_filenames, \
+                    #     'Several nse or ntt files have the same unit_id!!!'
+
+                    self.nse_ntt_filenames[uid] = filename
```
self.nse_ntt_filenames[( chan_name, chan_id)] = filename

samuelgarcia commented on neo/rawio/neuralynxrawio.py:
```diff
@@ -133,26 +151,36 @@ def _parse_header(self):
+                    else:
+                        data = np.memmap(filename, dtype=dtype, mode='r', offset=HEADER_SIZE)
+
+                    self._spike_memmap[uid] = data
```
OK

samuelgarcia commented on neo/rawio/neuralynxrawio.py:
```diff
@@ -133,26 +151,36 @@ def _parse_header(self):
                     for unit_id in unit_ids:
                         # a spike channel for each (chan_id, unit_id)
-                        self.internal_unit_ids.append((chan_id, unit_id))
+                        self.internal_unit_ids.append((uid, unit_id))
```
This could disappear.

samuelgarcia commented on neo/rawio/neuralynxrawio.py:
```diff
@@ -133,26 +151,36 @@ def _parse_header(self):
+                        self.internal_unit_ids.append((uid, unit_id))
 
-                        unit_name = "ch{}#{}".format(chan_id, unit_id)
+                        unit_name = "ch{}#{}".format(uid, unit_id)
```
Let s go back to unit_name = "ch{}#{}".format(chan_id, unit_id) or to
unit_name = "{}ch{}#{}".format(chan_name, chan_id, unit_id)
so the user can make the link with the sofwtare

samuelgarcia commented:
I made some comment in the code.

Best.

sin-mike commented:
@samuelgarcia 

Ok, I tried with tuple ids (chan

but...

it breaks at header info formatting:
`206         sig_channels = np.array(sig_channels, dtype=_signal_channel_dtype)`

The signal channel dtypes come from:
`from .baserawio import (BaseRawIO, _signal_channel_dtype,   _unit_channel_dtype, _event_channel_dtype)`

So, in my version, it was just:
`sig_channels.append((chan_name, uid, info['sampling_rate'], 'int16', units, gain, offset, group_id))`

Then, with tuple ids, I must write: 
`sig_channels.append((chan_name, (chan_name, chan_id), info['sampling_rate'], 'int16', units, gain, offset, group_id))`

with tuples, it breaks the header dtypes, and I tough the header is something built in stone, so I'm not allowed to change its fields.

If I'm allowed to change the header structure, then, no problem, tuple ids will work, otherwise, `id` field must be ints

samuelgarcia commented:
(chan_name, chan_id) would not be for the header['signal_channels'] but for self.ncs_filenames[ (chan_name, chan_id)]
